transports/quic/src/muxer.rs

```diff
@@ -0,0 +1,381 @@
+        Poll::Ready(Ok(substream))
+    }
+
+    fn poll_close(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        let Inner {
+            substreams,
+            connection,
+            ..
+        } = &mut *self.inner.lock();
+        if connection.is_drained() {
+            return Poll::Ready(Ok(()));
+        }
+
+        for substream in substreams.keys() {
+            let _ = connection.finish_substream(*substream);
+        }
+
+        loop {
+            if connection.send_stream_count() == 0 && !connection.is_closed() {
+                connection.close()
+            }
+            if let ConnectionEvent::ConnectionLost(_) = ready!(connection.poll_event(cx)) {
+                return Poll::Ready(Ok(()));
+            }
+        }
+    }
+}
+
+/// State of a single substream.
+#[derive(Debug, Default, Clone)]
+struct SubstreamState {
+    /// Waker to wake if the substream becomes readable or stopped.
+    read_waker: Option<Waker>,
+    /// Waker to wake if the substream becomes writable or stopped.
+    write_waker: Option<Waker>,
+    /// Waker to wake if the substream becomes closed or stopped.
+    finished_waker: Option<Waker>,
+
+    is_write_closed: bool,
+}
+
+impl SubstreamState {
+    fn wake_all(&mut self) {
+        if let Some(waker) = self.read_waker.take() {
+            waker.wake();
+        }
+        if let Some(waker) = self.write_waker.take() {
+            waker.wake();
+        }
+        if let Some(waker) = self.finished_waker.take() {
+            waker.wake();
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct Substream {
+    id: quinn_proto::StreamId,
+    muxer: Arc<Mutex<Inner>>,
+}
+
+impl Substream {
+    fn new(id: quinn_proto::StreamId, muxer: Arc<Mutex<Inner>>) -> Self {
+        Self { id, muxer }
+    }
+}
+
+impl AsyncRead for Substream {
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        mut buf: &mut [u8],
+    ) -> Poll<io::Result<usize>> {
+        let mut muxer = self.muxer.lock();
+
+        let mut stream = muxer.connection.recv_stream(self.id);
+        let mut chunks = match stream.read(true) {
+            Ok(chunks) => chunks,
+            Err(quinn_proto::ReadableError::UnknownStream) => {
+                return Poll::Ready(Ok(0));
+            }
+            Err(quinn_proto::ReadableError::IllegalOrderedRead) => {
+                unreachable!(
+                    "Illegal ordered read can only happen if `stream.read(false)` is used."
+                );
+            }
+        };
+        let mut bytes = 0;
+        let mut pending = false;
+        loop {
+            let chunk = match chunks.next(buf.len()) {
+                Ok(Some(chunk)) if !chunk.bytes.is_empty() => chunk,
+                Ok(_) => break,
+                Err(err @ quinn_proto::ReadError::Reset(_)) => {
+                    return Poll::Ready(Err(io::Error::new(io::ErrorKind::ConnectionReset, err)))
```

Comment:
Should we call `buf.clear()` here? Connection resets typically mean that anything we received can be considered garbage.

Reply:
We are transparent about the connection reset, so I would leave that up to the application / caller to decide what to do with the bytes received thus far.